File: haskell/acronym/src/Acronym.hs
module Acronym (abbreviate) where

import qualified Data.Text as T
import qualified Data.Char as Ch

abbreviate :: String -> String
abbreviate xs = (getOnlyUppers . splitInWords) textToAbreviate 
    where 
        textToAbreviate = T.pack xs
        splitInWords text = 
            T.split (`elem`  " -_") text
        getOnlyUppers textList = 
            T.unpack $ T.concat $ map getLetters textList
        getLetters text
            | text == T.empty = text
            | T.all Ch.isUpper text = 
                T.singleton $ Ch.toUpper $ T.head text 
            | otherwise = 
                T.cons 
                    (Ch.toUpper $ T.head text) 
                    (T.filter Ch.isUpper (T.tail text))   
        

File: fsharp/acronym/Acronym.fs
module Acronym

let regex (pat: string) (str: string) =
    System.Text.RegularExpressions.Regex.IsMatch(str, pat)

let abbreviate (phrase: string) =

    phrase.Split [| ' '; '\t'; '-'; '\n' |]
    |> Seq.map (String.filter System.Char.IsLetter)
    |> Seq.filter (regex "^\S")
    |> Seq.map (Seq.head >> System.Char.ToUpper >> string)
    |> String.concat ""
